src/stream/src/executor/source/source_backfill_executor.rs

```diff
@@ -503,21 +504,28 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
                                 last_barrier_time = Instant::now();
 
                                 if self_paused {
-                                    backfill_stream = select_with_strategy(
-                                        input.by_ref().map(Either::Left),
-                                        paused_reader.take().expect("no paused reader to resume"),
-                                        select_strategy,
-                                    );
+                                    // command_paused has a higher priority.
+                                    if !command_paused {
+                                        backfill_stream = select_with_strategy(
+                                            input.by_ref().map(Either::Left),
+                                            paused_reader
+                                                .take()
+                                                .expect("no paused reader to resume"),
+                                            select_strategy,
+                                        );
+                                    }
                                     self_paused = false;
                                 }
 
                                 let mut split_changed = false;
                                 if let Some(ref mutation) = barrier.mutation.as_deref() {
                                     match mutation {
                                         Mutation::Pause => {
+                                            command_paused = true;
                                             pause_reader!();
```

Comment:
I think we should also ensure `pause_reader!()` should only be called when only one of `command_paused` and `self_paused` is false. Otherwise, the right arm of the backfill stream will be lost. See [pause_reader](https://github.com/risingwavelabs/risingwave/blob/d856dd9df1bf498860d15ae633a90ec44135d441/src/stream/src/executor/source/source_backfill_executor.rs#L402) implementation.

Reply:
If source backfilling is expected to strictly adhere to the barrier pause, it must pause here.
I'll look into `pause_reader!`.

Reply:
Got your idea. Fixed.